lite/examples/sound_classification/ios/SoundClassification/GMLAudioRecord/GMLAudioRecord.m

```diff
@@ -0,0 +1,185 @@
+
+                    switch (converterStatus) {
+                      case AVAudioConverterOutputStatus_HaveData: {
+                        GMLFloatBuffer *floatBuffer =
+                            [[GMLFloatBuffer alloc] initWithData:pcmBuffer.floatChannelData[0]
+                                                            size:pcmBuffer.frameLength];
+
+                        NSError *frameBufferFormatError = nil;
+                        NSError *loadError = nil;
+
+                        if (pcmBuffer.frameLength == 0) {
+                          [GMLUtils createCustomError:&frameBufferFormatError
+                                             withCode:GMLAudioErrorCodeInvalidArgumentError
+                                          description:@"You may have to try with a different "
+                                                      @"channel count or sample rate"];
+                          completionHandler(nil, frameBufferFormatError);
+                        } else if ((pcmBuffer.frameLength % recordingFormat.channelCount) != 0) {
+                          [GMLUtils
+                              createCustomError:&frameBufferFormatError
+                                       withCode:GMLAudioErrorCodeInvalidArgumentError
+                                    description:
+                                        @"You have passed an unsupported number of channels."];
+                          completionHandler(nil, frameBufferFormatError);
+                        } else if (![self->_ringBuffer loadWithBuffer:floatBuffer
+                                                               offset:0
+                                                                 size:floatBuffer.size
+                                                                error:&loadError]) {
+                          completionHandler(nil, loadError);
+                        } else {
+                          GMLFloatBuffer *outFloatBuffer = [self->_ringBuffer.buffer copy];
+                          completionHandler(outFloatBuffer, nil);
+                        }
+                        break;
+                      }
+                      case AVAudioConverterOutputStatus_Error: {
+                        completionHandler(nil, conversionError);
+                        break;
+                      }
+                      default:
+                        completionHandler(nil, nil);
+                        break;
+                    }
+                  });
+                }];
+
+  NSError *engineStartError = nil;
+
+  [_audioEngine prepare];
+  [_audioEngine startAndReturnError:&engineStartError];
+
+  if (engineStartError) {
+    completionHandler(nil, engineStartError);
+  }
+}
+
+- (void)checkPermissionsAndStartTappingMicrophoneWithCompletionHandler:
+    (void (^)(GMLFloatBuffer *_Nullable buffer, NSError *_Nullable error))completionHandler {
+  [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted) {
```

Comment:
Please do permission check inside startTappingMicrophoneWithCompletionHandler so that we have only one "start" method. We should always check if the app has permission before starting the audio recording.